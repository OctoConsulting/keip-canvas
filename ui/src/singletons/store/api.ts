import { Edge } from "reactflow"
import { EipFlowNode, Layout, RouterKey } from "../../api/flow"
import { Attributes } from "../../api/generated/eipFlow"
import { EipId } from "../../api/id"

export interface EipConfig {
  attributes: Attributes
  children: string[]
  eipId: EipId
  description?: string
  routerKey?: RouterKey
}

export interface AppStore {
  nodes: EipFlowNode[]
  edges: Edge[]
  eipConfigs: Record<string, EipConfig>
  selectedChildNode: string | null
  layout: Layout
}

export interface SerializedFlow {
  nodes: AppStore["nodes"]
  edges: AppStore["edges"]
  eipConfigs: AppStore["eipConfigs"]
  version: string
}
