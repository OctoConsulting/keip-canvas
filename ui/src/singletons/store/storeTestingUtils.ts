import { renderHook } from "@testing-library/react"

// Zustand mock import (see '/home/aasaad/projects/keip-canvas/ui/__mocks__/zustand.ts')
// @ts-expect-error the reset function does not exist as a zustand export.
// It is added to the mock as a convenience for setup and cleanup of the store in the tests
import { resetStore } from "zustand"

// WARNING: In test files, import the reset method last, to ensure it's properly initialized.
export const resetMockStore = resetStore as (s: object) => void

export function renderAndUnwrapHook<T>(hookFn: () => T): T {
  const { result } = renderHook(hookFn)
  return result.current
}
