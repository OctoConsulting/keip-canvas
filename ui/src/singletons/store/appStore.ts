import isDeepEqual from "fast-deep-equal"
import { temporal } from "zundo"
import { create } from "zustand"
import { createJSONStorage, persist } from "zustand/middleware"
import debounce from "../../utils/debounce"

import { AppStore } from "./api"

const NO_PERSIST = new Set(["appActions", "reactFlowActions"])

// If app becomes too slow, might need to switch to async persistent storage.
export const useAppStore = create<AppStore>()(
  persist(
    temporal(
      (_) => ({
        nodes: [],
        edges: [],
        eipNodeConfigs: {},
        selectedChildNode: null,
        layout: {
          orientation: "horizontal",
          density: "comfortable",
        },
      }),
      {
        limit: 50,
        partialize: (state) => {
          const newNodes = state.nodes.map((node) => {
            const n = { ...node }
            const { selected, draggable, dragging, positionAbsolute, ...rest } =
              n
            return rest
          })

          const { eipNodeConfigs, edges, layout } = state
          return { eipNodeConfigs, layout, edges, nodes: newNodes }
        },

        equality: (pastState, currentState) =>
          isDeepEqual(pastState, currentState),

        handleSet: (handleSet) =>
          debounce<typeof handleSet>((state) => handleSet(state), 1000, true),
      }
    ),
    {
      name: "eipFlow",
      version: 0,
      storage: createJSONStorage(() => localStorage),
      partialize: (state) =>
        Object.fromEntries(
          Object.entries(state).filter(([key]) => !NO_PERSIST.has(key))
        ),
    }
  )
)
