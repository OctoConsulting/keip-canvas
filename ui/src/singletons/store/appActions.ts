// TODO: Bugfix. Nested updates of the 'eipNodeConfigs' store field are mutating
// references to the current state, this could lead to subtle bugs. Modify a deep
// copy of the state object and return that instead. Consider using Immer.

import { nanoid } from "nanoid/non-secure"
import { Edge, Position, XYPosition } from "reactflow"
import {
  ChannelMapping,
  DYNAMIC_EDGE_TYPE,
  DynamicEdge,
  EIP_NODE_TYPE,
  EipFlowNode,
  Layout,
} from "../../api/flow"
import { AttributeType } from "../../api/generated/eipComponentDef"
import { ChildNodeId, EipId, ROOT_PARENT } from "../../api/id"
import { newFlowLayout } from "../../components/layout/layouting"
import { AppStore, EipNodeConfig } from "./api"
import { useAppStore } from "./appStore"

export const createDroppedNode = (eipId: EipId, position: XYPosition) =>
  useAppStore.setState((state) => {
    const node = newNode(eipId, position, state.layout.orientation)
    return {
      nodes: [...state.nodes, node],
      eipNodeConfigs: {
        ...state.eipNodeConfigs,
        [node.id]: { attributes: {}, children: {} },
      },
    }
  })

export const updateNodeLabel = (id: string, label: string) => {
  let error: Error | undefined
  useAppStore.setState((state) => {
    const updatedNodes = state.nodes.map((node) => {
      if (node.id === id) {
        return { ...node, data: { ...node.data, label } }
      } else {
        if (node.data.label === label) {
          error = new Error("Node labels must be unique")
        }
        return node
      }
    })
    return { nodes: error ? state.nodes : updatedNodes }
  })
  return error
}

export const updateNodeDescription = (id: string, description: string) =>
  useAppStore.setState((state) => {
    const configs = { ...state.eipNodeConfigs }
    configs[id].description = description
    return { eipNodeConfigs: configs }
  })

export const updateEipAttribute = (
  id: string,
  parentId: string,
  attrName: string,
  value: AttributeType
) =>
  useAppStore.setState((state) => {
    const configs = { ...state.eipNodeConfigs }
    if (parentId === ROOT_PARENT) {
      configs[id].attributes[attrName] = value
    } else {
      const children = configs[parentId].children
      configs[parentId].children = {
        ...children,
        [id]: {
          ...children[id],
          attributes: {
            ...children[id].attributes,
            [attrName]: value,
          },
        },
      }
    }
    return { eipNodeConfigs: configs }
  })

export const deleteEipAttribute = (
  id: string,
  parentId: string,
  attrName: string
) =>
  useAppStore.setState((state) => {
    const nodeConfigs = state.eipNodeConfigs
    if (parentId === ROOT_PARENT) {
      const updatedAttrs = { ...nodeConfigs[id].attributes }
      delete updatedAttrs[attrName]
      return {
        eipNodeConfigs: {
          ...nodeConfigs,
          [id]: { ...nodeConfigs[id], attributes: updatedAttrs },
        },
      }
    } else {
      const updatedAttrs = {
        ...nodeConfigs[parentId].children[id].attributes,
      }
      delete updatedAttrs[attrName]
      return {
        eipNodeConfigs: {
          ...nodeConfigs,
          [parentId]: {
            ...nodeConfigs[parentId],
            children: {
              ...nodeConfigs[parentId].children,
              [id]: {
                ...nodeConfigs[parentId].children[id],
                attributes: updatedAttrs,
              },
            },
          },
        },
      }
    }
  })

// TODO: Ensure no more than one default mapping edge can be set
export const updateDynamicEdgeMapping = (
  edgeId: string,
  mapping: Partial<ChannelMapping>
) =>
  useAppStore.setState((state) => ({
    edges: state.edges.map((edge) => {
      if (edge.id === edgeId) {
        const dynamic = validateDynamicEdgeType(edge)
        return {
          ...dynamic,
          data: {
            ...dynamic.data,
            mapping: { ...dynamic.data?.mapping, ...mapping },
          },
        }
      }
      return edge
    }),
  }))

export const updateContentRouterKey = (
  nodeId: string,
  keyName: string,
  attrName: string,
  value: AttributeType
) =>
  useAppStore.setState((state) => ({
    eipNodeConfigs: {
      ...state.eipNodeConfigs,
      [nodeId]: {
        ...state.eipNodeConfigs[nodeId],
        routerKey: {
          name: keyName,
          attributes: {
            ...state.eipNodeConfigs[nodeId].routerKey?.attributes,
            [attrName]: value,
          },
        },
      },
    },
  }))

export const updateEnabledChildren = (nodeId: string, children: string[]) =>
  useAppStore.setState((state) => {
    const configs = { ...state.eipNodeConfigs }
    configs[nodeId].children = children.reduce(
      (accum, name) => {
        accum[name] = { name }
        return accum
      },
      {} as EipNodeConfig["children"]
    )

    return { eipNodeConfigs: configs }
  })

export const updateSelectedChildNode = (childId: ChildNodeId) =>
  useAppStore.setState(() => ({ selectedChildNode: childId }))

export const clearSelectedChildNode = () =>
  useAppStore.setState(() => ({ selectedChildNode: null }))

export const clearFlow = () =>
  useAppStore.setState(() => ({
    nodes: [],
    edges: [],
    eipNodeConfigs: {},
    selectedChildNode: null,
  }))

export const clearDiagramSelections = () =>
  useAppStore.setState((state) => ({
    nodes: state.nodes.map((node) => ({ ...node, selected: false })),
    edges: state.edges.map((edge) => ({ ...edge, selected: false })),
  }))

// TODO: Should importing throw an error on failure instead?
export const importFlowFromJson = (json: string) =>
  useAppStore.setState(() => {
    const imported = JSON.parse(json) as Partial<AppStore>
    if (isStoreType(imported)) {
      return {
        nodes: imported.nodes,
        edges: imported.edges,
        eipNodeConfigs: imported.eipNodeConfigs,
      }
    }
    console.error("Failed to import an EIP flow JSON. Malformed input")
    return {}
  })

export const updateLayoutOrientation = (orientation: Layout["orientation"]) =>
  useAppStore.setState((state) => {
    const newLayout: Layout = {
      ...state.layout,
      orientation: orientation,
    }
    const nodes = newFlowLayout(state.nodes, state.edges, newLayout)
    return {
      nodes: nodes,
      layout: newLayout,
    }
  })

export const toggleLayoutDensity = () =>
  useAppStore.setState((state) => {
    const newDensity =
      state.layout.density === "compact" ? "comfortable" : "compact"
    const newLayout: Layout = {
      ...state.layout,
      density: newDensity,
    }
    const nodes = newFlowLayout(state.nodes, state.edges, newLayout)
    return {
      nodes: nodes,
      layout: newLayout,
    }
  })

const newNode = (
  eipId: EipId,
  position: XYPosition,
  orientation: Layout["orientation"]
) => {
  const id = nanoid(10)
  const isHorizontal = orientation === "horizontal"
  const node: EipFlowNode = {
    id: id,
    type: EIP_NODE_TYPE,
    position: position,
    targetPosition: isHorizontal ? Position.Left : Position.Top,
    sourcePosition: isHorizontal ? Position.Right : Position.Bottom,
    data: {
      eipId: eipId,
    },
  }
  return node
}

const validateDynamicEdgeType = (edge: Edge) => {
  if (edge.type !== DYNAMIC_EDGE_TYPE) {
    throw new Error(
      `The provided edge did not have the expected type: "${edge.type}". Should be "${DYNAMIC_EDGE_TYPE}"`
    )
  }
  return edge as DynamicEdge
}

const isStoreType = (state: unknown): state is AppStore => {
  const store = state as AppStore
  return (
    store.nodes !== undefined &&
    store.edges !== undefined &&
    store.eipNodeConfigs !== undefined
  )
}
