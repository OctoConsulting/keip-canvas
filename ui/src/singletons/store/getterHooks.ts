import { useShallow } from "zustand/react/shallow"
import { DYNAMIC_EDGE_TYPE, DynamicEdge } from "../../api/flow"
import { AppStore } from "./api"
import { useAppStore } from "./appStore"

export const useNodeCount = () => useAppStore((state) => state.nodes.length)

export const useGetNodes = () => useAppStore((state) => state.nodes)

export const useGetLayout = () => useAppStore((state) => state.layout)

export const useFlowStore = () =>
  useAppStore(
    useShallow((state: AppStore) => ({
      nodes: state.nodes,
      edges: state.edges,
    }))
  )

export const useUndoRedo = () =>
  useAppStore(() => ({
    undo: useAppStore.temporal.getState().undo,
    redo: useAppStore.temporal.getState().redo,
  }))

export const useSerializedStore = () =>
  useAppStore((state) =>
    JSON.stringify({
      nodes: state.nodes,
      edges: state.edges,
      eipNodeConfigs: state.eipConfigs,
    })
  )

export const useGetNodeDescription = (id: string) =>
  useAppStore((state) => state.eipConfigs[id]?.description)

export const useGetEipAttribute = (id: string, attrName: string) =>
  useAppStore((state) => state.eipConfigs[id].attributes[attrName])

export const useGetEnabledChildren = (id: string) =>
  useAppStore(
    useShallow((state) =>
      state.eipConfigs[id] ? state.eipConfigs[id].children : []
    )
  )

export const useGetSelectedChildNode = () =>
  useAppStore((state) => state.selectedChildNode)

export const useGetContentRouterKey = (nodeId: string) =>
  useAppStore(useShallow((state) => state.eipConfigs[nodeId]?.routerKey))

export const useGetRouterDefaultEdgeMapping = (routerId: string) =>
  useAppStore((state) =>
    state.edges.find(
      (edge) =>
        edge.source === routerId &&
        edge.type === DYNAMIC_EDGE_TYPE &&
        (edge as DynamicEdge).data?.mapping.isDefaultMapping
    )
  )
