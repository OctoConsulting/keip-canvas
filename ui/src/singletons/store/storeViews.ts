import { Edge } from "reactflow"
import { EipFlowNode, Layout } from "../../api/flow"
import { EipId } from "../../api/generated/eipFlow"
import { useAppStore } from "./appStore"

// Warning: the following exports provide non-reactive access to the store's state
export const getNodesView = (): readonly EipFlowNode[] =>
  useAppStore.getState().nodes

export const getEdgesView = (): readonly Edge[] => useAppStore.getState().edges

export const getLayoutView = (): Readonly<Layout> =>
  useAppStore.getState().layout

export const getEipId = (nodeId: string): Readonly<EipId> | undefined =>
  useAppStore.getState().eipConfigs[nodeId]?.eipId
