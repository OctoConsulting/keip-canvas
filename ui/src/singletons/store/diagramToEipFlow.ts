import isDeepEqual from "fast-deep-equal"
import { useStoreWithEqualityFn } from "zustand/traditional"
import {
  DYNAMIC_EDGE_TYPE,
  DynamicEdge,
  EipFlowNode,
  EipNodeData,
  RouterKey,
} from "../../api/flow"
import {
  Attributes,
  EipChildNode,
  EipFlow,
  EipNode,
  FlowEdge,
} from "../../api/generated/eipFlow"
import { EipId } from "../../api/generated/eipFlow"
import {
  CHANNEL_ATTR_NAME,
  DEFAULT_OUTPUT_CHANNEL_NAME,
  lookupContentBasedRouterKeys,
  lookupEipComponent,
} from "../eipDefinitions"
import { AppStore, EipConfig } from "./api"
import { useAppStore } from "./appStore"
import { getEipId } from "./storeViews"

const EIP_NAMESPACE_TO_XML_PREFIX: Record<string, string> = {
  xml: "int-xml",
  "web-services": "ws",
}

export const useEipFlow = () =>
  useStoreWithEqualityFn(
    useAppStore,
    (state) => diagramToEipFlow(state),
    isDeepEqual
  )

// TODO: Dynamic router logic is over-complicating this method, extract to
// a separate method and apply modifications after building original flow.
const diagramToEipFlow = (state: AppStore): EipFlow => {
  const nodeLookup = createNodeLookupMap(state.nodes)

  const routerChildMap = new Map<string, EipChildNode[]>()
  const routerAttrMap = new Map<string, Attributes>()

  const edges: FlowEdge[] = state.edges.map((edge) => {
    const source = nodeLookup.get(edge.source)?.label || edge.source
    const target = nodeLookup.get(edge.target)?.label || edge.target
    const edgeId = `ch-${source}-${target}`

    if (edge.type === DYNAMIC_EDGE_TYPE) {
      addRouterChannelMapping(edgeId, edge, routerChildMap, routerAttrMap)
    }

    return {
      id: edgeId,
      source,
      target,
      type: edge.sourceHandle === "discard" ? "discard" : "default",
    }
  })

  const nodes: EipNode[] = state.nodes.map((node) => {
    const eipId = getEipId(node.id)!
    const eipComponent = lookupEipComponent(eipId)!
    const namespace = eipId.namespace

    const routerKey = state.eipConfigs[node.id].routerKey
    const routerKeyAttrs =
      eipComponent.connectionType === "content_based_router" && routerKey
        ? getRouterKeyAttributes(eipId, routerKey)
        : {}

    return {
      id: node.data.label ? node.data.label : node.id,
      eipId: {
        ...eipId,
        namespace: EIP_NAMESPACE_TO_XML_PREFIX[namespace] ?? namespace,
      },
      description: state.eipConfigs[node.id]?.description,
      role: eipComponent.role,
      connectionType: eipComponent.connectionType,
      attributes: {
        ...state.eipConfigs[node.id]?.attributes,
        ...routerKeyAttrs.attributes,
        ...routerAttrMap.get(node.id),
      },
      children: buildChildTree(node.id, state.eipConfigs).concat(
        routerKeyAttrs.child ?? [],
        routerChildMap.get(node.id) ?? []
      ),
    }
  })

  return { nodes, edges }
}

// TODO: Refactor. Ugly.
const buildChildTree = (
  rootId: string,
  eipConfigs: AppStore["eipConfigs"]
): EipChildNode[] => {
  const top = eipConfigs[rootId].children.map((childId) =>
    childConfigToNode(eipConfigs[childId])
  )

  const queue = [...top]

  while (queue.length > 0) {
    const curr = queue.shift()!
    if (curr.childIds && curr.childIds.length > 1) {
      curr.children = curr.childIds.map((childId) =>
        childConfigToNode(eipConfigs[childId])
      )
      queue.push(...curr.children)
    }
    delete curr.childIds
  }

  return top
}

const childConfigToNode = (config: EipConfig) =>
  ({
    name: config.eipId.name,
    attributes: config.attributes,
    childIds: config.children,
  }) as EipChildNode & {
    childIds?: string[]
  }

const createNodeLookupMap = (nodes: EipFlowNode[]) => {
  const map = new Map<string, EipNodeData>()
  nodes.forEach((node) => map.set(node.id, node.data))
  return map
}

// TODO: consider extracting the ChannelMapping and RouterKey logic to the XML translation backend
const addRouterChannelMapping = (
  channelId: string,
  edge: DynamicEdge,
  routerChildMap: Map<string, EipChildNode[]>,
  routerAttributeMap: Map<string, Attributes>
) => {
  const mapping = edge.data?.mapping
  if (!mapping) {
    return
  }

  if (mapping.isDefaultMapping) {
    routerAttributeMap.set(edge.source, {
      [DEFAULT_OUTPUT_CHANNEL_NAME]: channelId,
    })
  } else {
    const child: EipChildNode = {
      name: mapping.mapperName,
      attributes: {
        [CHANNEL_ATTR_NAME]: channelId,
      },
    }
    if (mapping.matcherValue) {
      child.attributes![mapping.matcher.name] = mapping.matcherValue
    }
    const curr = routerChildMap.get(edge.source) ?? []
    routerChildMap.set(edge.source, curr.concat(child))
  }
}

const getRouterKeyAttributes = (eipId: EipId, routerKey: RouterKey) => {
  const routerKeyDef = lookupContentBasedRouterKeys(eipId)
  if (!routerKeyDef) {
    return {}
  }
  switch (routerKeyDef.type) {
    case "attribute": {
      return { attributes: filterEmptyAttrs(routerKey.attributes ?? {}) }
    }
    case "child": {
      return {
        child: {
          name: routerKey.name,
          attributes: filterEmptyAttrs(routerKey.attributes ?? {}),
        },
      }
    }
  }
}

const filterEmptyAttrs = (attrs: Attributes) =>
  Object.keys(attrs).reduce((acc, key) => {
    const val = attrs[key]
    if (val) {
      acc[key] = val
    }
    return acc
  }, {} as Attributes)
