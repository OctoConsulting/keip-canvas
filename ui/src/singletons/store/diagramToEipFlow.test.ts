import { act, renderHook } from "@testing-library/react"
import { expect, test, vi } from "vitest"
import { useEipFlow } from "../store"
import { resetMockStore } from "./storeTestingUtils"
import childBasedRouterFlow from "./testdata/store-initializers/childBasedRouterFlow.json"
import standardFlow from "./testdata/store-initializers/standardFlow.json"

vi.mock("zustand")

test("standard diagram to EipFlow success", () => {
  act(() => {
    resetMockStore(standardFlow)
  })

  const { result } = renderHook(() => useEipFlow())
  expect(result.current).toMatchSnapshot()
})

test("diagram includes a router with child based matcher to EipFlow success", () => {
  act(() => {
    resetMockStore(childBasedRouterFlow)
  })

  const { result } = renderHook(() => useEipFlow())
  expect(result.current).toMatchSnapshot()
})
