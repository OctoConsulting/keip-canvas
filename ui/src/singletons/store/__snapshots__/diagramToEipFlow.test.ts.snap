// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`diagram includes a router with child based matcher to EipFlow success 1`] = `
{
  "edges": [
    {
      "id": "ch-rn79OHQyAI-cGUxaVOQ7L",
      "source": "rn79OHQyAI",
      "target": "cGUxaVOQ7L",
      "type": "default",
    },
    {
      "id": "ch-cGUxaVOQ7L-1su4DIeR80",
      "source": "cGUxaVOQ7L",
      "target": "1su4DIeR80",
      "type": "default",
    },
    {
      "id": "ch-cGUxaVOQ7L-RJlDxzUi2a",
      "source": "cGUxaVOQ7L",
      "target": "RJlDxzUi2a",
      "type": "default",
    },
  ],
  "nodes": [
    {
      "attributes": {},
      "children": [],
      "connectionType": "source",
      "description": undefined,
      "eipId": {
        "name": "inbound-channel-adapter",
        "namespace": "ftp",
      },
      "id": "rn79OHQyAI",
      "role": "endpoint",
    },
    {
      "attributes": {},
      "children": [
        {
          "attributes": {
            "expression": "/target/protocol",
          },
          "name": "xpath-expression",
        },
        {
          "attributes": {
            "channel": "ch-cGUxaVOQ7L-1su4DIeR80",
            "value": "jms",
          },
          "name": "mapping",
        },
        {
          "attributes": {
            "channel": "ch-cGUxaVOQ7L-RJlDxzUi2a",
            "value": "kafka",
          },
          "name": "mapping",
        },
      ],
      "connectionType": "content_based_router",
      "description": undefined,
      "eipId": {
        "name": "xpath-router",
        "namespace": "int-xml",
      },
      "id": "cGUxaVOQ7L",
      "role": "router",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": undefined,
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "kafka",
      },
      "id": "RJlDxzUi2a",
      "role": "endpoint",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": undefined,
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "jms",
      },
      "id": "1su4DIeR80",
      "role": "endpoint",
    },
  ],
}
`;

exports[`standard diagram to EipFlow success 1`] = `
{
  "edges": [
    {
      "id": "ch-inbound-test-router",
      "source": "inbound",
      "target": "test-router",
      "type": "default",
    },
    {
      "id": "ch-test-router-httpOut",
      "source": "test-router",
      "target": "httpOut",
      "type": "default",
    },
    {
      "id": "ch-test-router-test-filter",
      "source": "test-router",
      "target": "test-filter",
      "type": "default",
    },
    {
      "id": "ch-test-filter-fileOut",
      "source": "test-filter",
      "target": "fileOut",
      "type": "default",
    },
    {
      "id": "ch-test-filter-MZ1rIWIK3s",
      "source": "test-filter",
      "target": "MZ1rIWIK3s",
      "type": "discard",
    },
  ],
  "nodes": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {
            "fixed-rate": "2000",
          },
          "name": "poller",
        },
      ],
      "connectionType": "source",
      "description": "message incoming",
      "eipId": {
        "name": "inbound-channel-adapter",
        "namespace": "integration",
      },
      "id": "inbound",
      "role": "endpoint",
    },
    {
      "attributes": {
        "default-output-channel": "ch-test-router-httpOut",
        "expression": "headers.protocol",
        "phase": "init",
      },
      "children": [
        {
          "attributes": {
            "channel": "ch-test-router-test-filter",
            "value": "file",
          },
          "name": "mapping",
        },
      ],
      "connectionType": "content_based_router",
      "description": undefined,
      "eipId": {
        "name": "router",
        "namespace": "integration",
      },
      "id": "test-router",
      "role": "router",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": "send an http message",
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "http",
      },
      "id": "httpOut",
      "role": "endpoint",
    },
    {
      "attributes": {
        "expression": "headers.filename != null",
      },
      "children": [],
      "connectionType": "tee",
      "description": undefined,
      "eipId": {
        "name": "filter",
        "namespace": "integration",
      },
      "id": "test-filter",
      "role": "router",
    },
    {
      "attributes": {},
      "children": [],
      "connectionType": "sink",
      "description": "write to file",
      "eipId": {
        "name": "outbound-channel-adapter",
        "namespace": "file",
      },
      "id": "fileOut",
      "role": "endpoint",
    },
    {
      "attributes": {
        "level": "WARN",
      },
      "children": [],
      "connectionType": "sink",
      "description": undefined,
      "eipId": {
        "name": "logging-channel-adapter",
        "namespace": "integration",
      },
      "id": "MZ1rIWIK3s",
      "role": "endpoint",
    },
  ],
}
`;
